File: simulate.m
function sim_file=simulate(data_file, prefix, sim_vertices, sim_woi,...
    sim_signals, dipole_orientations, dipole_moments, sim_patch_sizes, SNR, spm_path)

addpath(spm_path);

% Start SPM
spm('defaults','eeg');
spm_jobman('initcfg');

Dmesh=spm_eeg_load(data_file);

%% get location to simulate dipole on this mesh
%simpos=Dmesh.inv{1}.mesh.tess_mni.vert(sim_vertices,:); 

% Make signal have unit variance
%sim_signals=sim_signals./repmat(std(sim_signals'),size(sim_signals,2),1)';

[Dnew,~]=spm_eeg_simulate({Dmesh}, prefix, sim_vertices, sim_signals,...
    dipole_orientations, sim_woi, [], SNR,...
    [], [], sim_patch_sizes, dipole_moments, []);

[a1, ~, ~]=fileparts(data_file);
sim_file=fullfile(a1, Dnew.fname);

File: simulate.py
import matlab.engine
import numpy as np

from util import get_spm_path


def run_current_density_simulation(data_file, prefix, sim_vertices, sim_signals, dipole_moments, sim_patch_sizes, SNR,
                                   sim_woi=[-np.inf, np.inf], mat_eng=None):
    spm_path = get_spm_path()

    if np.isscalar(sim_vertices):
        sim_vertices=[sim_vertices]
    sim_vertices=[x+1 for x in sim_vertices]
    if np.isscalar(dipole_moments):
        dipole_moments=[dipole_moments]
    if np.isscalar(sim_patch_sizes):
        sim_patch_sizes=[sim_patch_sizes]

    close_matlab = False
    if mat_eng is None:
        mat_eng = matlab.engine.start_matlab()
        close_matlab = True

    sim_fname=mat_eng.simulate(
        data_file,
        prefix,
        matlab.double(sim_vertices),
        matlab.double(sim_woi),
        matlab.double(sim_signals.tolist()),
        matlab.double([]),
        matlab.double(dipole_moments),
        matlab.double(sim_patch_sizes),
        float(SNR),
        spm_path,
        nargout=1
    )

    if close_matlab:
        mat_eng.close()

    return sim_fname


def run_dipole_simulation(data_file, prefix, sim_vertices, sim_signals, dipole_orientations, dipole_moments, sim_patch_sizes,
                          SNR, sim_woi=[-np.inf, np.inf], mat_eng=None):
    spm_path = get_spm_path()

    if np.isscalar(sim_vertices):
        sim_vertices=[sim_vertices]
    sim_vertices = [x + 1 for x in sim_vertices]
    if np.isscalar(dipole_moments):
        dipole_moments=[dipole_moments]
    if np.isscalar(sim_patch_sizes):
        sim_patch_sizes=[sim_patch_sizes]

    close_matlab = False
    if mat_eng is None:
        mat_eng = matlab.engine.start_matlab()
        close_matlab = True

    sim_fname=mat_eng.simulate(
        data_file,
        prefix,
        matlab.double(sim_vertices),
        matlab.double(sim_woi),
        matlab.double(sim_signals.tolist()),
        matlab.double(dipole_orientations.tolist()),
        matlab.double(dipole_moments),
        matlab.double(sim_patch_sizes),
        float(SNR),
        spm_path,
        nargout=1
    )

    if close_matlab:
        mat_eng.close()

    return sim_fname